### PR TITLE
Run npm in a new Java process

### PR DESCRIPTION
**Pull request by yahavi**
Build info code: https://github.com/jfrog/build-info/pull/286
See example here: https://github.com/jfrog/project-examples/pull/190

https://github.com/jfrog/jenkins-artifactory-plugin/issues/147

### CODE REVIEW

File: src/main/java/org/jfrog/hudson/pipeline/common/executors/MavenGradleEnvExtractor.java [hunk 1]
```diff
@@ -3,95 +3,23 @@
...
-
-    private Deployer publisher;
-    private Resolver resolver;
-    private Run build;
-    private BuildInfo buildInfo;
-    private TaskListener buildListener;
-    private Launcher launcher;
-    private FilePath tempDir;
-    private EnvVars env;
+public class MavenGradleEnvExtractor extends EnvExtractor {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/pipeline/common/executors/MavenGradleEnvExtractor.java]: Do we still need this class? Can we remove it and stay with *EnvExtractor* only?

File: src/main/java/org/jfrog/hudson/pipeline/common/executors/NpmEnvExtractor.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        this.args = args;
+        this.path = path;
+    }
+
+    @Override
+    protected void addExtraConfiguration(ArtifactoryClientConfiguration configuration) {
+        ArtifactoryClientConfiguration.NpmHandler npmHandler = configuration.npmHandler;
+        npmHandler.setNpmExecutablePath(npmExe);
+        npmHandler.setNpmInstallArgs(args);
+        npmHandler.setNpmPath(path);
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/pipeline/common/executors/NpmEnvExtractor.java]: The above can be simplified the below (have the first line removed):
```
        configuration.npmHandler.setNpmExecutablePath(npmExe);
        configuration.npmHandler.setNpmInstallArgs(args);
        configuration.npmHandler.setNpmPath(path);
 
```

File: src/main/java/org/jfrog/hudson/pipeline/common/executors/NpmExecutor.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    String copyExtractorJars(FilePath tempDir) throws IOException, InterruptedException {
+        File extractorJar = PluginDependencyHelper.getExtractorJar(env);
+        FilePath dependencyDir = PluginDependencyHelper.getActualDependencyDirectory(extractorJar, tempDir);
+        String absoluteDependencyDirPath = dependencyDir.getRemote();
+        return absoluteDependencyDirPath.replace("\\", "/");
+    }
+
+    ArgumentListBuilder getArgs(String absoluteDependencyDirPath, String classToExecute) {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/pipeline/common/executors/NpmExecutor.java]: I think this method can be private.

File: src/main/java/org/jfrog/hudson/pipeline/common/executors/NpmExecutor.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                buildInfo, deployer, resolver, listener, launcher, tempDir, env, args, path, npmExe);
+        envExtractor.execute();
+        String absoluteDependencyDirPath = copyExtractorJars(tempDir);
+        Utils.launch("npm", launcher, getArgs(absoluteDependencyDirPath, classToExecute), env, listener, ws);
+        String generatedBuildPath = env.get(BuildInfoFields.GENERATED_BUILD_INFO);
+        buildInfo.append(Utils.getGeneratedBuildInfo(build, listener, launcher, generatedBuildPath));
+        buildInfo.setAgentName(Utils.getAgentName(ws));
+    }
+
+    String copyExtractorJars(FilePath tempDir) throws IOException, InterruptedException {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/pipeline/common/executors/NpmExecutor.java]: I think this method can be private.

File: pom.xml [hunk 5]
```diff
@@ -323,6 +323,11 @@
             <artifactId>commons-io</artifactId>
             <version>2.5</version>
         </dependency>
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-compress</artifactId>
```
Comment (eyalbe4) [inline pom.xml]: Was commons-compress added intentionally?

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmInstallStep.java [hunk 6]
```diff
@@ -63,6 +64,11 @@ public void setResolverId(String resolverId) {
         this.resolverId = resolverId;
     }
 
+    @DataBoundSetter
+    public void setJavaArgs(String javaArgs) {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmInstallStep.java]: I suggest adding the following comment:
This setter has been added to allow java remote debugging.

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmPublishStep.java [hunk 7]
```diff
@@ -63,6 +64,11 @@ public void setDeployerId(String deployerId) {
         this.deployerId = deployerId;
     }
 
+    @DataBoundSetter
+    public void setJavaArgs(String javaArgs) {
```
Comment (eyalbe4) [inline src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmPublishStep.java]: I suggest adding the following comment:
This setter has been added to allow java remote debugging.

File: src/main/java/org/jfrog/hudson/pipeline/common/executors/MavenGradleEnvExtractor.java (hunk 1 above)
Comment (yahavi) [reply src/main/java/org/jfrog/hudson/pipeline/common/executors/MavenGradleEnvExtractor.java]: We need it for `addExtraConfiguration`

File: pom.xml (hunk 5 above)
Comment (yahavi) [reply pom.xml]: Yes. In the build info code, we're using decompression: https://github.com/jfrog/build-info/blob/master/build-info-extractor-npm/src/main/java/org/jfrog/build/extractor/npm/extractor/NpmPublish.java#L99
Without it I'm getting this error:
```
Error: Unable to initialize main class org.jfrog.build.extractor.npm.extractor.NpmPublish
Caused by: java.lang.NoClassDefFoundError: org/apache/commons/compress/compressors/gzip/GzipCompressorInputStream
```

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmInstallStep.java (hunk 6 above)
Comment (yahavi) [reply src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmInstallStep.java]: I added the comment in the javaArgs variable above

File: src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmPublishStep.java (hunk 7 above)
Comment (yahavi) [reply src/main/java/org/jfrog/hudson/pipeline/declarative/steps/npm/NpmPublishStep.java]: I added the comment in the javaArgs variable above

**Pull request merged**